### ISSUE TITLE
scala.MatchError: 3099.0 (of class java.lang.Float)

### ISSUE THREAD

**Issue by salonegupta** [elastic/elasticsearch-hadoop#1044]
### What kind an issue is this?

- [x] Bug report.

### Issue description

While fetching any nested field from Spark ES connector, the field is being converted into an array which is leading to the above-mentioned error.

### Steps to reproduce

I am using Spark 2.11 with Java 8 and Elasticsearch 5.4. I have a nested document structure which is as follows:
```
{
  "mappings": {
    "event": {
      "properties": {
        "userId": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "event": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "properties": {
          "properties": {
            "action": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "label": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "revenue": {
              "type": "float"
            }
          }
        },
        "receivedAt": {
          "type": "date"
        }
      }
    }
  }
}
```
I am trying to fetch `event`, `userId` and `properties.revenue` fields from ES using Spark ES library. When I try to execute following code

Code:

```
SparkConf conf = new SparkConf().setAppName("es_test").setMaster("local[2]");

JavaSparkContext sc = new JavaSparkContext(conf);
SQLContext sqlContext = new SQLContext(sc);

Map<String, String> config = new java.util.HashMap<>();
config.put("es.read.field.include", "properties.revenue, event, userId");

Dataset<Row> ds = JavaEsSparkSQL.esDF(sqlContext, "events/timeline", config);
Dataset df = ds.select("userId", "event", "properties.revenue").filter("event = 'Completed Order'"); 

df.foreach((ForeachFunction<Row>) row -> {
    System.out.println(row);
});
```
following schema tree is getting created for the Dataset which is being created here:
```
(java.lang.String) root
|-- userId: string (nullable = true)
|-- event: string (nullable = true)
|-- revenue: array (nullable = true)
|    |-- element: float (containsNull = true)
```
Here, revenue is being treated as an array whereas as per ES schema, it is defined as float. And, it throws following error when above mentioned code is being executed:

Strack trace:

```
scala.MatchError: 3099.0 (of class java.lang.Float)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$ArrayConverter.toCatalystImpl(CatalystTypeConverters.scala:160)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$ArrayConverter.toCatalystImpl(CatalystTypeConverters.scala:154)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:241)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$StructConverter.toCatalystImpl(CatalystTypeConverters.scala:231)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
at org.apache.spark.sql.catalyst.CatalystTypeConverters$$anonfun$createToCatalystConverter$2.apply(CatalystTypeConverters.scala:379)
at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:61)
at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$3.apply(ExistingRDD.scala:58)
at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
at scala.collection.Iterator$class.foreach(Iterator.scala:893)
at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
at org.apache.spark.rdd.RDD$$anonfun$foreach$1$$anonfun$apply$28.apply(RDD.scala:918)
at org.apache.spark.rdd.RDD$$anonfun$foreach$1$$anonfun$apply$28.apply(RDD.scala:918)
at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2062)
at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2062)
at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
at org.apache.spark.scheduler.Task.run(Task.scala:108)
at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
``` 
In case of String field, the error is `scala.MatchError: R1234 (of class java.lang.String)`

### Version Info

OS:         :  OS X Yosemite
JVM         :  Oracle Java 8
Hadoop/Spark:  2.11
ES-Hadoop   :  6.0.0
ES          :  5.4

**Comment by PowerToThePeople111** [comment]
I have the very same problem. I have nested fields which I included into `es.read.field.as.array.include`. Those fields then have the problem of not getting matched correctly.

I use:
ES-Spark: 6.2.1
JVM: Oracle Java 8
Spark: 2.1.1 (with Scala 2.11)
ES: 5.2.0

**Comment by jbaiera** [comment]
Going through open issues right now and cleaning them up. I tried to reproduce this but could not get it to fail:

```scala
    // INDEX
    val index = "sparksql-struct-index"
    val typeName = "doc"
    val indexAndType = index + "/" + typeName

    try {
      RestUtils.delete(index)
    } catch {
      case _: Throwable => // what ever
    }

    // CREATE INDEX
    RestUtils.touch(index)

    // EXISTING MAPPINGS
    val mapping =
      s"""
         |{
         |  "properties": {
         |    "userId": {
         |      "type": "text"
         |    },
         |    "event": {
         |      "type": "text"
         |    },
         |    "properties": {
         |      "properties": {
         |        "action": {
         |          "type": "text"
         |        },
         |        "label": {
         |          "type": "text"
         |        },
         |        "revenue": {
         |          "type": "float"
         |        }
         |      }
         |    },
         |    "receivedAt": {
         |      "type": "date"
         |    }
         |  }
         |}
      """.stripMargin
    RestUtils.putMapping(index, typeName, mapping.getBytes())

    // VALIDATE MAPPING
    val resolvedMapping = RestUtils.getMapping(indexAndType)
    println(resolvedMapping)
    println()
    val flattened: Map[String, FieldType] = resolvedMapping.flatten()
    flattened.foreach(println(_))
    println()

    // EXISTING DATA
    RestUtils.postData(indexAndType,
      """
        |{
        |  "userId": "me",
        |  "event": "HTTP Get",
        |  "properties": {
        |    "action": "Get",
        |    "label": "Spark",
        |    "revenue": 3099.0
        |  },
        |  "receivedAt": "2018-03-17T00:54:03Z"
        |}
      """.stripMargin.getBytes())
    RestUtils.refresh(index)

    // VALIDATE WRITE
    println("Written Data:")
    val validateWrite = RestUtils.get(indexAndType+"/_search")
    println(validateWrite)

    // READ DATA (SQL)
    val readConf = Map("es.read.field.include" -> "properties.revenue, event, userId")
    val df = sql.read.format("es").options(readConf).load(indexAndType)

    df.printSchema()

    df.show(20)
    df.select("userId", "event", "properties.revenue").show()
```
Outputs were:
```
...
doc=[event=TEXT, properties=[action=TEXT, label=TEXT, revenue=FLOAT], receivedAt=DATE, userId=TEXT]

(event,TEXT)
(properties,OBJECT)
(properties.action,TEXT)
(properties.label,TEXT)
(properties.revenue,FLOAT)
(receivedAt,DATE)
(userId,TEXT)

Written Data:
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"skipped":0,"failed":0},"hits":{"total":1,"max_score":1.0,"hits":[{"_index":"sparksql-struct-index","_type":"doc","_id":"G42hMmIBP9JsBLrawPuT","_score":1.0,"_source":
{
  "userId": "me",
  "event": "HTTP Get",
  "properties": {
    "action": "Get",
    "label": "Spark",
    "revenue": 3099.0
  },
  "receivedAt": "2018-03-17T00:54:03Z"
}
      }]}}
...
root
 |-- event: string (nullable = true)
 |-- properties: struct (nullable = true)
 |    |-- revenue: float (nullable = true)
 |-- userId: string (nullable = true)
...
+--------+----------+------+
|   event|properties|userId|
+--------+----------+------+
|HTTP Get|  [3099.0]|    me|
+--------+----------+------+
...
+------+--------+-------+
|userId|   event|revenue|
+------+--------+-------+
|    me|HTTP Get| 3099.0|
+------+--------+-------+
...
```
If I'm missing anything in this reproduction, please let me know. I will update it and see about re-opening the issue. Until then, closing.

**Comment by PowerToThePeople111** [comment]
Thanks jbaiera for your help!

The issue as we found out is: the problematic field is of type object so we can add data specific for this special row. So basically, for the given type in the elastic index this object field can have different properties.

Using esRDD works fine, since it stores the content of the doc as a simple string. All other schema inferring functions to persist the data fail (esDF, load(index), ...). I fear this is because of the varying length of the list: when i query elastic for only one row it is actually returned fine by all functions I tested.

A schema merge should in general be possible as long as the object-field's properties are always of the same type. And in the case of conflict, they could be at least imported as string imho.
